Add NTRIP Client implementation

hect95 opened the pull request:
This PR introduces an NTRIP client implementation that manages connections between an NTRIP server and a Novatel GNSS receiver. Key features include:

- **NTRIP Client**: Connects to an NTRIP server to request and transmit RTCMv3 corrections.
- **GNSS Receiver Integration**: Handles communication with a Novatel GNSS receiver, including configuration and data parsing.
- **Data Handling**: Supports parsing of RTCM and NMEA data, and sends valid NMEA sentences to the NTRIP server.
- **Logging and Configuration**: Configurable logging levels and settings, with support for fixed GPS locations.
- **Docker Suppor**t: Added `dockerfile`, `dev.sh`, and `run.sh` scripts for containerised deployment.

hect95 committed:
Add ntrip client script

hect95 committed:
Add gnss tcp scocket and comm functions

hect95 committed:
WIP: Sending RTCMv3 to GNSS

hect95 committed:
WIP: Add better logging; Use fixed gnss data

hect95 committed:
Add nmea_generator class for debugging

hect95 committed:
Sending all fixed nmea sentences

hect95 committed:
Send all nmea sentences from GNSS to ntrip server

hect95 committed:
Add runtime flags

- Add --use-fix-location to choose fixed NMEA data to sent to the
   NTRIP server rather than data from the GNSS
- Add --debug to choos debug logging mode

hect95 committed:
Add novatel client to configure icom ports

hect95 committed:
Send only GPGGA data to the NTRIP server

- Changing default gnss_port to 3003

hect95 committed:
Remove unused function and add comments

hect95 committed:
It works!

hect95 committed:
Adding dev docker and fix comments

hect95 committed:
Add pre-commit

- Add and fix pre-commit linting suggestions and errors

hect95 committed:
Configure GNSS port automatically

hect95 committed:
Update pre-commit file

hect95 committed:
Reconnect to NTRIP server if connection is lost

hect95 committed:
Remove unnecessary operations

- Remove chunk_size parameter

hect95 committed:
Avoid sending invalid NMEA data to NTRIP server

hect95 committed:
Move novatel_client script to debug dir

hect95 committed:
Log when waiting for valid NMEA data

hect95 committed:
Add Docker build and push workflow

GreatAlexander commented on scripts/params.yaml:
```diff
@@ -0,0 +1,10 @@
+ntrip_port: 9301
+mountpoint: 'MSM_iMAX'
+username: '40017819B'
+password: 'UOFE002'
```
These should not be uploaded in a git commit, instead they should be a git secret.

hect95 replied:
I think we can retrive these params from the repo secrets every time we run the script.

hect95 replied:
We will need a Github Token configured in the server. I am okay with this

GreatAlexander replied:
Does that mean we need internet access every time we run the script? Or are secrets somehow cloned locally?
In any case this repo needs internet to work, so it's ok in either case

hect95 replied:
We need internet :)

hect95 replied:
As we discussed, I've set up the [av_ntrip_credentials](https://github.com/ipab-rad/av_ntrip_credentials) repository for the NTRIP server login details. This repo is now included as a submodule, so only those with the right permissions can clone it.

GreatAlexander commented on dev.sh:
Best to keep the dev/runtime scripts up to date with improvements made in other repos, e.g. for reference https://github.com/ipab-rad/av_ouster/blob/main/dev.sh

hect95 replied:
Done

GreatAlexander commented on run.sh:
Best to keep the dev/runtime scripts up to date with improvements made in other repos, e.g. for reference https://github.com/ipab-rad/av_ouster/blob/main/runtime.sh

Also, plz rename `runtime.sh` for consistency

hect95 replied:
I updated it and renamed it to `runtime.sh`

GreatAlexander commented on dockerfile:
```diff
@@ -0,0 +1,45 @@
+# Use the official Ubuntu 22.04 as a base image
+FROM ubuntu:22.04 AS base
```
For any new images, could we try starting with `24.04`? We should eventually update all our images like so.

hect95 replied:
Updated to `24.04`

GreatAlexander commented on scripts/debug/novatel_client.py:
```diff
@@ -0,0 +1,124 @@
+    parser.add_argument(
+        '--port',
+        type=int,
+        default=3003,
```
Default port in comment does not match default in code.

GreatAlexander commented on scripts/debug/novatel_client.py:
```diff
@@ -0,0 +1,124 @@
+        help='GNSS port number to connect and configure',
+    )
+
+    # Parse the arguments
```
Unnecessary comment

hect95 replied:
`novatel_client.py` removed

GreatAlexander commented on scripts/debug/novatel_client.py:
```diff
@@ -0,0 +1,124 @@
+    # Parse the arguments
+    args = parser.parse_args()
+
+    # Access the port number
```
Unnecessary comment

hect95 replied:
Done

hect95 replied:
`novatel_client.py` removed

GreatAlexander commented on scripts/debug/novatel_client.py:
```diff
@@ -0,0 +1,124 @@
+    # Access the port number
+    port_number = args.port
+
+    # Configure logging
+    formatter = colorlog.ColoredFormatter(
+        '%(asctime)s - %(log_color)s%(levelname)s%(reset)s - %(message)s',
+        datefmt='%Y-%m-%d %H:%M:%S',
+    )
+
+    console_handler = colorlog.StreamHandler()
+    console_handler.setFormatter(formatter)
+    logging.basicConfig(level=logging.INFO, handlers=[console_handler])
```
This configuration should not be in main (Specially since variables like `formatter` and `console_handler` are only needed to configure `logging`.
I would put all this in a separate method which is called in main, which then would look something like `logging = configureLogging()` in main.

hect95 replied:
`novatel_client.py` removed

GreatAlexander commented on scripts/debug/novatel_client.py:
```diff
@@ -0,0 +1,124 @@
+    logging.basicConfig(level=logging.INFO, handlers=[console_handler])
+
+    # Replace with your GNSS receiver's IP and port
+    client = NovatelClient(gnss_ip='172.31.0.90', gnss_port=port_number)
```
`gnss_ip` should be an argument and not hard-coded here.

hect95 replied:
As dicussed, we do not need this script anymore. I will remove it.

GreatAlexander commented on scripts/debug/novatel_client.py:
```diff
@@ -0,0 +1,124 @@
+        while True:
+            try:
+                # Receive the initial message to identify the ICOM port name
+                msg = client.gnss_socket.recv(1024)
```
Magic number? This should be a constant somewhere with a clear name to what this is for.

hect95 replied:
`novatel_client.py` removed

GreatAlexander commented on scripts/debug/novatel_client.py:
```diff
@@ -0,0 +1,124 @@
+
+
+if __name__ == '__main__':
+    # Initialize the argument parser
```
Unnecessary comment

hect95 replied:
Done

hect95 replied:
`novatel_client.py` removed

GreatAlexander commented on scripts/debug/novatel_client.py:
```diff
@@ -0,0 +1,124 @@
+            )
+
+            # Read the response to confirm configuration
+            response = self.gnss_socket.recv(1024).decode('utf-8')
```
Magic number? This should be a constant somewhere with a clear name to what this is for.

hect95 replied:
`ntrip_client.py` removed

GreatAlexander commented on scripts/nmea_generator.py:
```diff
@@ -0,0 +1,126 @@
+            f'{self.generate_rmc_sentence()}\r\n'
+            f'{self.generate_gst_sentence()}\r\n'
+        )
+        return sentences
```
Do we really need/consume the three sentences GGA/RMC/GST?

hect95 replied:
We actually don't. I will remove this function

hect95 replied:
Done

GreatAlexander reviewed (commented):
Also another comment: Should we add any sort of ROS2 node/monitoring on this? Or are we relying on the diagnostics of e.g. Novatel to track the performance of this ntrip cllient?

GreatAlexander commented on scripts/ntrip_client.py:
```diff
@@ -0,0 +1,539 @@
+            time.sleep(1)
+            continue
+
+        # Configure GNSS
```
Comments like these are not needed

hect95 replied:
All unnecessary comments are removed

GreatAlexander commented on scripts/ntrip_client.py:
```diff
@@ -0,0 +1,539 @@
+        self.ntrip_socket.send(request.encode())
+
+        try:
+            response = self.ntrip_socket.recv(1024).decode('ISO-8859-1')
```
Magic number? This should be a constant somewhere with a clear name to what this is for.

Do this for all other magic numbers!

hect95 replied:
Done

GreatAlexander commented on scripts/ntrip_client.py:
```diff
@@ -0,0 +1,539 @@
+        debug_mode=args.debug,
+    )
+
+    # Run client
```
For example, I think all comments in this main are not needed / self evident (e.g. `#Run client` for `client.run()` )

hect95 replied:
All unnecessary comments are removed

hect95 committed:
Remove ntrip_client.py

- Remove get_fix_nmea_sentences() from NMEA generator

hect95 committed:
Upgrade docker to ubuntu 24.04

hect95 commented:
> Also another comment: Should we add any sort of ROS2 node/monitoring on this? Or are we relying on the diagnostics of e.g. Novatel to track the performance of this ntrip cllient?

As discussed, we won't port this script to a ROS2 node, as we will rely on log files instead.

hect95 commented:
> Consider saving logging prints to a log file for debugging

TODO

GreatAlexander commented on scripts/nmea_generator.py:
So who uses this script and for what? Where are these methods being used?
Now that the main function that depended on the sentence generation methods was removed, I don't know what this script is for.
Is it for local testing?
If so, could we make an automated test as part of the CI that uses the script to test the ntrip client functionality? This can be a future issue/PR

hect95 replied:
`NtripClient` uses it to generate and validate NMEA data based on the provided fix GPS location.

The sentence generation was a function to merge `GPGGA`, `GPRMC` and `GPGST` sentences. We currently only use `GPGGA` so that merging was not necessary anymore. I used it before to understand NTRIP server requests etc. I think we only need `GPGGA` data to receive corrections from the server. 

- I will delete `generate_rmc_sentence()` and`generate_gst_sentence`

hect95 replied:
Done

GreatAlexander commented on run.sh:
```diff
@@ -0,0 +1,14 @@
+#!/bin/bash
+
+# This scripts runs runtime ntrip docker container and
+#  passes any arguments to the ntrip_client app
```
"ntrip_client app" App? Do you mean the python script?

Perhaps a bit confusing use of "app" since we actually _do_ have and use ntrip client apps on our phones, so I'd recommend calling it Python script or code

hect95 replied:
`run.sh` has been updated and renamed to `runtime.sh`. This new version contains a different comment

hect95 committed:
Update docker bash scripts

- run.sh is renamed to runtime.sh

hect95 committed:
Remove unused functions

hect95 committed:
Fix code as per comments

- Remove unnecessary comments
- Avoid magic numbers
- Convert non essential INFO logs to DEBUG

hect95 committed:
Use private NTRIP credentials

- Remove params.yaml containing sensitive log in details
- Add av_ntrip_credentials private repository as submodule
- Use smartnet_novatel_params.yaml from av_ntrip_credentials dir
- Add example_params.yaml for guidance

hect95 committed:
Improve logging

- Log to a timestamped file
- Improve logs msgs
- Parse Novatel GNSS binary response to extract msg directly

hect95 commented:
> Consider saving logging prints to a log file for debugging

@GreatAlexander I have configured `logging` to save logs into a `<time_stamp>_ntrip_client.log` file

hect95 committed:
Support private submodules in Docker build

hect95 committed:
Update README

hect95 committed:
Fix pre-commit warnings

GreatAlexander commented on scripts/ntrip_client.py:
```diff
@@ -0,0 +1,588 @@
+        try:
+            self.ntrip_socket.send(request.encode())
+            self.nmea_request_sent = True
+            # logging.debug('NMEA message sent to NTRIP server.')
```
Commented code

hect95 replied:
Removed

GreatAlexander commented on scripts/ntrip_client.py:
```diff
@@ -0,0 +1,588 @@
+                if nmea_sentence:
+                    # logging.debug(
+                    #     f'NMEA data received from GNSS: {nmea_sentence}'
+                    # )
```
Commented code

hect95 replied:
Removed

hect95 committed:
Remove commented code

hect95 committed:
Rename dockerfile

GreatAlexander commented:
Ship it! :ship: :shipit: :rocket:
